### PR TITLE
FINERACT-2081: Enable inline COB execution for locked loans

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

If the loan is locked, but there is any error attached to the lock, it means the prior execution of COB / Inline COB failed. 

For user experience it would be better if we let the inline COB to be triggered again in these situations hence noone is processing the loan at the moment, it just indicates there was some problem during the last processing.

[FINERACT-2081](https://issues.apache.org/jira/browse/FINERACT-2081)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/cob/service/InlineLoanCOBExecutorServiceImpl.java [hunk 1]
```diff
@@ -241,7 +241,7 @@ private boolean isLockOverrulable(LoanAccountLock loanAccountLock) {
         if (isBypassUser()) {
             return true;
         } else {
-            return StringUtils.isNotBlank(loanAccountLock.getError());
+            return StringUtils.isBlank(loanAccountLock.getError());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/service/InlineLoanCOBExecutorServiceImpl.java]: I dont think it is correct to do this. Please undo this change. We can change the lock if we are going with bypass user or there was an error stored on the lock.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java [hunk 2]
```diff
@@ -75,7 +77,9 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
                     proceed(filterChain, request, response);
                 } else {
                     try {
+                        log.info("calculateRelevantLoanIds ==>");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java]: No info log please! Use debug!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java [hunk 3]
```diff
@@ -240,7 +250,8 @@ private Long getTopLevelLoanIdFromBatchRequest(BatchRequest batchRequest) throws
 
     private List<Long> getLoanIdsFromApi(String pathInfo) {
         List<Long> loanIds = getLoanIdList(pathInfo);
-        if (isLoanHardLocked(loanIds)) {
+        log.info("getLoanIdsFromApi {}", loanIds.size());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java]: No info log please!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java [hunk 4]
```diff
@@ -217,7 +227,7 @@ private List<Long> getLoanIdsFromBatchApi(BodyCachingHttpServletRequestWrapper r
             // check the body for Loan ID
             Long loanId = getTopLevelLoanIdFromBatchRequest(batchRequest);
             if (loanId != null) {
-                if (isLoanHardLocked(loanId)) {
+                if (isLoanHardLocked(loanId) && isLockOverrulable(loanId)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java]: Wrong condition. If the lock has an error stored, we should allow to run the inline cob!

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java [hunk 5]
```diff
@@ -240,7 +250,8 @@ private Long getTopLevelLoanIdFromBatchRequest(BatchRequest batchRequest) throws
 
     private List<Long> getLoanIdsFromApi(String pathInfo) {
         List<Long> loanIds = getLoanIdList(pathInfo);
-        if (isLoanHardLocked(loanIds)) {
+        log.info("getLoanIdsFromApi {}", loanIds.size());
+        if (isLoanHardLocked(loanIds) && isLockOverrulable(loanIds)) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java]: Wrong condition. If the lock has an error stored, we should allow to run the inline cob!

File: fineract-provider/src/main/java/org/apache/fineract/cob/service/InlineLoanCOBExecutorServiceImpl.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/cob/service/InlineLoanCOBExecutorServiceImpl.java]: Done! Change undone

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java (hunk 2 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java]: Done! Logs removed

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBFilterHelper.java]: Done! Logs removed

**Review by alberto-art3ch — COMMENTED** [review]
PR Updated

File: fineract-e2e-tests-core/src/test/java/org/apache/fineract/test/stepdef/common/BatchApiStepDef.java [hunk 6]
```diff
@@ -506,16 +505,13 @@ public void runBatchApiCreateAndApproveLoanRescheduleWithGivenUser(String fromDa
...
-    @When("Batch API call with created user and the following data results a {int} error and a {string} error message:")
-    public void runBatchApiCreateAndApproveLoanRescheduleWithGivenUserLockedByCobError(int errorCodeExpected, String errorMessageType,
+    @When("Batch API call with created user and the following data results a {int} statuscode and a {string} error message:")
+    public void runBatchApiCreateAndApproveLoanRescheduleWithGivenUserLockedByCobError(int httpCodeExpected, String errorMessageType,
             DataTable table) throws IOException {
         String idempotencyKey = UUID.randomUUID().toString();
         Response<PostLoansResponse> loanResponse = testContext().get(TestContextKey.LOAN_CREATE_RESPONSE);
         Long loanId = loanResponse.body().getLoanId();
 
-        LoanRescheduleErrorMessage loanRescheduleErrorMessage = LoanRescheduleErrorMessage.valueOf(errorMessageType);
```
Comment (adamsaghy) [inline fineract-e2e-tests-core/src/test/java/org/apache/fineract/test/stepdef/common/BatchApiStepDef.java]: Why to remove these checks?

File: fineract-e2e-tests-core/src/test/java/org/apache/fineract/test/stepdef/common/BatchApiStepDef.java [hunk 7]
```diff
@@ -542,17 +538,10 @@ public void runBatchApiCreateAndApproveLoanRescheduleWithGivenUserLockedByCobErr
         Boolean isEnclosingTransaction = Boolean.valueOf(enclosingTransaction);
         Response<List<BatchResponse>> batchResponseList = batchApiApi.handleBatchRequests(requestList, isEnclosingTransaction, headerMap)
                 .execute();
-        String errorToString = batchResponseList.errorBody().string();
```
Comment (adamsaghy) [inline fineract-e2e-tests-core/src/test/java/org/apache/fineract/test/stepdef/common/BatchApiStepDef.java]: Why to remove these checks?

File: fineract-e2e-tests-runner/src/test/resources/features/LoanReschedule.feature [hunk 8]
```diff
@@ -658,6 +658,7 @@ Feature: LoanReschedule
     When Batch API call with created user and with steps: rescheduleLoan from "16 January 2024" to "31 January 2024" submitted on date: "10 January 2024", approveReschedule on date: "10 January 2024" runs with enclosingTransaction: "true"
     Then Admin checks that last closed business date of loan is "09 January 2024"
 
+  @temp
```
Comment (adamsaghy) [inline fineract-e2e-tests-runner/src/test/resources/features/LoanReschedule.feature]: Do we need this new tag?

**Review by adamsaghy — APPROVED** [review]
LGTM
